### PR TITLE
compaction compare pact state tool

### PR DESCRIPTION
**Pull request by chessai**
compares pact state pre- and post-compaction

---
- To see the specific tasks where the Asana app for GitHub is being used, see below:
  - https://app.asana.com/0/0/1205378300255963

### CODE REVIEW

File: src/Chainweb/Pact/Backend/PactState.hs [hunk 1]
```diff
@@ -0,0 +1,337 @@
...
+    -- the chain ids at the current (version,height) pair
+    latestBlockHeight <- C.withDefaultLogger Error $ \logger -> do
+      let resetDb = False
+      withSqliteDb (unsafeChainId 0) logger cfg.pactDbDir resetDb $ \(SQLiteEnv db _) -> do
+        getLatestBlockHeight db
+    pure $ List.sort $ F.toList $ chainIdsAt cfg.chainwebVersion latestBlockHeight
+
+  -- Compaction is destructive. In order to do a streaming diff, we need to make
+  -- a copy of the pact state.
+  withSystemTempDirectory "pact-compact-diff" $ \tmpPactDbDir -> do
```
Comment (chessai) [inline src/Chainweb/Pact/Backend/PactState.hs:283]: there should be an argument for where the copy should go. and there should be a check that it's different from the pactDbDir. this way the post-compaction db can be inspected.

File: tools/cwtool/CwTool.hs [hunk 2]
```diff
@@ -104,7 +105,11 @@ topLevelCommands =
   , CommandSpec
       "compact"
       "Compact pact database"
-      compactMain
+      Chainweb.Pact.Backend.Compaction.main
+  , CommandSpec
+      "pact-db-compare"
```
Comment (chessai) [inline tools/cwtool/CwTool.hs:110]: this should probably be something like "compact-compare"

**Review by ak3n — APPROVED** [review]
LGTM 👍 

I think it would be nice to have a comment why exactly we use `Stream`s here, at first glance they look more like an overcomplication and it's unclear if we benefit from them.

**Comment by chessai** [comment]
> LGTM 👍 
> 
> I think it would be nice to have a comment why exactly we use `Stream`s here, at first glance they look more like an overcomplication and it's unclear if we benefit from them.

Sounds good to me. I'll add a note in the code. It's primarily because the memory requirements can get very high for all the combined tables.

File: src/Chainweb/Pact/Backend/PactState.hs [hunk 3]
```diff
@@ -0,0 +1,345 @@
...
+        withSqliteDb cid logger cfg.pactDbDir resetDb $ \(SQLiteEnv db _) -> do
+          diffEmptyRef <- newIORef True
+
+          let diff = diffLatestPactState (getLatestPactState db) (getLatestPactState cpDb)
+          flip S.mapM_ diff $ \utd -> do
+            writeIORef diffEmptyRef False
+            loggerFunIO logger Warn $ toLogMessage $
+              TextLog $ Text.decodeUtf8 $ BSL.toStrict $ Aeson.encode utd
+
+          diffEmpty <- readIORef diffEmptyRef
```
Comment (chessai) [inline src/Chainweb/Pact/Backend/PactState.hs:304]: This should be something that happens at the very end. We should have an `IORef (Map ChainId Bool)`.

File: src/Chainweb/Pact/Backend/PactState.hs [hunk 4]
```diff
@@ -0,0 +1,345 @@
...
+          when (not diffEmpty) $ do
+            loggerFunIO logger Warn $ toLogMessage $
```
Comment (chessai) [inline src/Chainweb/Pact/Backend/PactState.hs:306]: We should be logging to a file, not to stdout.

**Pull request merged**